### PR TITLE
Backport #14368 to v3.5

### LINKED ISSUE #16250: etcd 集群链接 报错context canceled
### Bug report criteria

- [ ] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [X] This is not a support request, support requests should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

etcd集群链接报错，集群链接配置 RejectOldCluster=true检测 checkVersion 有一个链接成功就跳过其它节点链接，直接cancal取消上下文，导致链接报错。
image

### What did you expect to happen?

集群链接报错
image

### How can we reproduce it (as minimally and precisely as possible)?

在本机环境，关闭etcd集群，然后启动，启动完成，立即启动运行etcd链接代码。注：在第一次运行的时候会报错，后面在重试启动未报错。
注：重复复现流程，关闭etcd集群在启动etcd集群，然后运行etcd链接代码。
##etcd链接代码
image

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
etcd Version: 3.5.9
Git SHA: bdbbde9
Go Version: go1.19.10
Go OS/Arch: linux/amd64

$ etcdctl version
etcdctl version: 3.5.9
API version: 3.5
```

</details>

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

### Relevant log output

```Shell
{"level":"warn","ts":"2023-07-15T11:32:45.366+0800","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retry
ing of unary invoker failed","target":"etcd-endpoints://0xc000203340/192.168.7.75:2379","attempt":0,"error":"rpc error: code = Cance
led desc = context canceled"}
{"level":"warn","ts":"2023-07-15T11:32:45.366+0800","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000203340/192.168.7.75:2379","attempt":0,"error":"rpc error
ing of unary invoker failed","target":"etcd-endpoints://0xc000203340/192.168.7.75:2379","attempt":0,"error":"rpc error: code = Cance
led desc = context canceled"}
```

### PR DESCRIPTION
**Pull request by CaojiamingAlan**
Fix #16250

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Review by liangyuanpeng — APPROVED** [review]

**Review by fuweid — APPROVED** [review]
LGTM

File: client/v3/client.go [hunk 1]
```diff
@@ -502,15 +502,15 @@ func (c *Client) checkVersion() (err error) {
 					return
 				}
 			}
-			if maj < 3 || (maj == 3 && min < 2) {
+			if maj < 3 || (maj == 3 && min < 4) {
```
Comment (ahrtr) [inline client/v3/client.go:505]: We should avoid hard code the version in source code. For stable releases (e.g. 3.5 and 3.4), it's fine. For the main branch, let's raise a ticket to programmatically get the supported versions.

Note we only support two releases, which are 3.5 and 3.4 for now. FYI. https://etcd.io/docs/v3.5/op-guide/versioning/

**Review by ahrtr — APPROVED** [review]

**Comment by ahrtr** [comment]
Please raise a followup ticket per https://github.com/etcd-io/etcd/pull/16270#discussion_r1267770143, before we merge this PR.

File: client/v3/client.go (hunk 1 above)
Comment (CaojiamingAlan) [reply client/v3/client.go:505]: Can I interpret is as we support the current version and one previous version only? 
i.e. for v3.6, v3.5 is valid, and for v3.7, v3.6 is still valid, while 3.5 will be out of date.

Comment (ahrtr) [reply client/v3/client.go:505]: > Can I interpret is as we support the current version and one previous version only?

YES. Currently we support 3.4 and 3.5. When 3.6.0 is formally released, then 3.4 will be out of support. When 3.7.0 is released, then 3.5 will be out of support.

Comment (CaojiamingAlan) [reply client/v3/client.go:505]: Raised #16273

Comment (ahrtr) [reply client/v3/client.go:505]: thx

**Pull request merged**